(MODULES-8419) Refactor to add support for GitLab

antaflos opened the pull request:
* Create a GitPlatformClient wrapper class with the methods required by
  pdksync main
* Keep the interface unchanged, meaning pdksync main uses the same
  methods as before, but now they are no longer Octokit-specific
* Create wrapper classes for Github and Gitlab API access, using
  octokit/octokit.rb for Github and NARKOZ/gitlab for Gitlab
* Create a wrapper class for pull request objects abstracting the
  differences between Github and Gitlab
* Keep Github and the Github API as defaults as before
* Extend configuration and constants for Git platform-specific settings
* Allow the user to configure which Git hosting platform to use
* Allow the user to configure the Git base URI under which to access
  Git repositories
* Set default Git base URI according to configured Git platform
* Add sanity checks to PdkSync::Constants and bail if GITHUB_TOKEN or
  GITLAB_TOKEN are unset, or if an unknown Git platform is
  configured
* Guard against ~/.pdksync.yml being an empty file or an empty YAML
  file. Both cases would result in trying to access non-existent keys,
  leading to failure.
* Constrain bundler to ~> 1.15 in pdksync.gemspec since bundler >= 2.0.0
  introduces some incompatibilities with older Rubygems versions.
* Update README on how to configure pdksync to use GitLab instead of
  GitHub

antaflos commented:
Not sure how to get Travis to set GITHUB_TOKEN here so that the spec tests can run. Please advise :)

antaflos commented:
Gently bumping this for visibility :)

eimlav commented:
Hey @antaflos . Thanks for contributing this PR it looks great! Apologies for the delay in reviewing your PR.

Up until now we have not received any external PRs so we have not modified our CI setup to take this into account. As a result, external PRs will fail in Travis. To remediate we will have to find a different solution to our current CI setup but for now I have checked out your PR locally and tested it myself. I encountered an issue installing the bundle which was caused by using bundler 2.0.1, so a constraint will need to be added in  `pdksync.gemspec` to pin bundler to `~> 1.15`. After doing so, all tests ran successfully for GitHub which is great to see! I setup a test GitLab repo and ran the tests using a GitLab token and they also ran successfully.

It would be helpful to explicitly state the config variables that need modified in order to use GitLab. As an example why, I found when testing that when I forgot to set `git_base_uri`, the GitHub URL was used even though I stated `git_platform` as `gitlab` resulting in errors. You could even add some minor logic to change the default `git_base_uri` to `https://gitlab.com` when `git_platform` is `gitlab` or just explicitly state in the README. Whatever you think is the best approach.

Having discussed with the team, we would be happy to merge these changes (after you have made the the previously stated updates) however going forward we would not be providing official support for the GitLab functionality that has been introduced. If you could add a note in `README.md` to state this it would be appreciated. Other than that awesome job. It'll be really nice to have this new functionality in pdksync! :)

Cheers and let me know if you have any questions.

antaflos commented:
Thank you for the feedback @eimlav!

Updates to the documentation (i.e. README) are still TODO but I will gladly incorporate them, along with addressing the other issues you mentioned. Expect an update to this PR in the next few days.

I have also opened https://tickets.puppetlabs.com/browse/MODULES-8419 a few weeks ago, in case you need a ticket to reference for your internal processes.

eimlav commented:
@antaflos Before you make your changes I should just point out that we use the ticket as part of the main commit message for example `(MODULES-8419) Refactor to add support for GitLab` so if you could amend this too that would be great. Thanks.

antaflos commented:
@eimlav I have updated the PR with your suggested changes and extended the README. I have also updated the commit message and the PR description and comment.

Please have a look when you get around to it.

eimlav commented:
Hey @antaflos. I've ran the acceptance tests on both GitHub and GitLab successfully. I've also manually tested the rake tasks and they all appear to function correctly. I'm going to go ahead now and merge the PR. Thank you for making this great PR!
